hc_dpki/src/seed.rs

```diff
@@ -0,0 +1,558 @@
+
+        let seed_data_deserialized: bundle::ReturnBundleData =
+            json::decode(&seed_data_string).unwrap();
+        let mut seed_data = SecBuf::with_secure(32);
+
+        util::pw_dec(
+            &seed_data_deserialized,
+            &mut passphrase,
+            &mut seed_data,
+            config,
+        )?;
+
+        match bundle.bundle_type.as_ref() {
+            "hcRootSeed" => Ok(FromBundle::Rs(RootSeed::new(seed_data))),
+            "hcDeviceSeed" => Ok(FromBundle::Ds(DeviceSeed::new(seed_data))),
+            "hcDevicePinSeed" => Ok(FromBundle::Dps(DevicePinSeed::new(seed_data))),
+            _ => Err(HolochainError::new(&format!("Invalid Bundle type"))),
+        }
+    }
+
+    ///  generate a persistence bundle with hint info
+    ///
+    ///  @param {string} passphrase - the encryption passphrase
+    ///
+    ///  @param {string} hint - additional info / description for persistence
+    ///
+    /// @return {KeyBundle} - bundle of the seed
+    pub fn get_seed_bundle(
+        &mut self,
+        passphrase: String,
+        hint: String,
+        config: Option<PwHashConfig>,
+    ) -> Result<bundle::KeyBundle, HolochainError> {
+        let mut passphrase = SecBuf::with_insecure_from_string(passphrase);
+        let seed_data: bundle::ReturnBundleData =
+            util::pw_enc(&mut self.seed_buf, &mut passphrase, config)?;
+
+        // convert -> to string -> to base64
+        let seed_data_serialized = json::encode(&seed_data).unwrap();
+        let seed_data_encoded = base64::encode(&seed_data_serialized);
+
+        Ok(bundle::KeyBundle {
+            bundle_type: self.seed_type.clone(),
+            hint,
+            data: seed_data_encoded,
+        })
+    }
+
+    ///  Initialize this seed class with persistence bundle type and private seed
+    ///
+    ///  @param {string} stype - the persistence bundle type
+    ///
+    ///  @param {SecBuf|string} seed - the private seed data (as a buffer or mnemonic)
+    pub fn new(stype: &String, sm: InitializeSeed) -> Self {
+        match sm {
+            SeedInit(s) => Seed {
+                seed_type: stype.clone(),
+                seed_buf: s,
+            },
+            MnemonicInit(phrase) => {
+                let mnemonic = Mnemonic::from_phrase(phrase, Language::English).unwrap();
+                let entropy: &[u8] = mnemonic.entropy();
+                let mut buf = SecBuf::with_insecure(entropy.len());
+                util::convert_array_to_secbuf(entropy, &mut buf);
+                Seed {
+                    seed_type: stype.clone(),
+                    seed_buf: buf,
+                }
+            }
+        }
+    }
+
+    /// Generated a mnemonic for the seed.
+    pub fn get_mnemonic(&mut self) -> Result<String, HolochainError> {
+        let entropy = self.seed_buf.read_lock();
+        let e = &*entropy;
+        let mnemonic = Mnemonic::from_entropy(e, Language::English).unwrap();
+        Ok(mnemonic.phrase().to_string())
```

Comment:
here again, we need some way of zeroing the internal memory used by mnemonic if possible... see if you can, or add a comment to come back to it.

Reply:
are you talking about the 'phrase' String stored within the Mnemonic struct? or maybe the entropy vec?

Reply:
I changed the return into a `into_phrase()`. Would that suffice?

Reply:
maybe? hard to know without looking into the internal memory management of the crate